### ISSUE TITLE
Tying parameters and feeding tensors to kernels

### ISSUE THREAD

**Issue by Bonnevie** [GPflow/GPflow#322]
It seems that any hyperparameters you feed a kernel on initialization are immediately enclosed in a `Param`object. So anything that's not a Numpy array will be rejected. 

How would we handle kernels where multiple sub-kernels share a parameter? If it doesn't break anything, it seems sensible to let the user define a `Param` instance before instantiating the sub-kernels, and then pass the `Param` to the kernels during initialization. Does this mess with AutoFlow or `tf_mode()` or `optimize()` seeing as the `Param` is no longer a child of the kernel? (I don't think it would, but I'm not sure)

More generally, I think we should allow users to set arbitrary tensors as kernel hyperparameters. This wouldn't really change much, as long as all free parameters are still encapsulated in `Param`. The only potential issue is breaking constraints, but I'd say that's a user-side problem - the constraints can already be broken by replacing the transform for instance. 

(This maybe also goes towards making the kernels more self-contained, which could let other libraries adopt them.)

**Comment by markvdw** [comment]
Not having the `Param` objects be children of the kernel would break the whole parameter system. It relies on being able to traverse the hierarchy of `Parameterized` objects, and getting each transformed variable into the free state vector. This currently works very well.

Arbitrary tensors can already be used as kernel hyperparameters. Just define a kernel that uses this correctly, right?

Sharing parameters could be implemented by implementing a new `Transform`, I think. It could be passed another `Param`. Then both transforms should take the same element from the free state. The details are more difficult though, and would need some thought on how to implement it.

**Comment by Bonnevie** [comment]
Won't it work if the `Param` is a child of a the model class `Model` containing the kernel? 

For the interested (this might be old news) you cannot use return statements in `__init__` but you can do something like this
```
class A:
    def __new__(cls, a):
        if a>0:
            return a
        else:
            return super(A, cls).__new__(cls)
             
    def __init__(self, a):
        self.a = a
```

**Comment by jameshensman** [comment]
Tying parameters would be a really useful feature. It would require real thought to implement effectively, but I think lots of people would want to use it. 

Even better, and slightly more general, would be the ability to define relationships between parameters. For example, something like

```python
m = GPflow.gpr.GPR(X, Y, ...)
m.kern.variance.set_to(m.likelihood.variance * 10)
```
which obviously includes tying parameters as a special case.

**Comment by Bonnevie** [comment]
@jameshensman: I think your example is general enough to cover my original request as well. 
This would also help with e.g. derivative kernels where your derivative kernel has to have the same parameters as the original kernel. 
Do you see any way forward for implementing this? I guess the key is to make the parameter's `_tf_array` attribute point to some other tensor quantity, and then make sure it doesn't take up any free_vars.

**Comment by jameshensman** [comment]
That sounds right, yes. I think it will be tricky, though, and there will
be lots of corner cases to catch.

**Comment by Bonnevie** [comment]
@jameshensman: I realize this is a sensitive proposition, but how essential is it that `Kern` subclasses parameterized and thus assumes one or more attributes is a `Param` or `Parameterized`? 
GPflow contains a number of what I'd call "agnostic" functions, that can work outside of the GPflow environment, such as the Kullback-Leibler functions and the conditionals. I'm thinking many people outside of the GPflow community would be happy if kernels were also agnostic in this sense. I think Dustin Tran from the Edward project argued as much when he dropped by a previous Issue.  
Right now, kernels will only work inside a `Model`, but as far as I can tell, replacing the `Param`s with any other numerical object would make the kernels agnostic in a heartbeat.

What we would lose:
1) a nice parameter hierarchy, e.g. being able to tell what parameters belong where when printing. 
2) automatic initialization of the kernels when used in a proper GPflow setting.  

Issue 1 is quite bad, but if we are e.g. tying parameters, then there no longer exists such a nice resolution of what parameters belong to which model anyway. Issue 2 would be a bit of a speed bump for most model initializations. 

One way we could get the best of both worlds would be to define the kernels without the `Parameterized`machinery, maybe split it off into a separate module, and then build GPflow-specific class wrappers that mix-in the `Parameterized` class and proper initialization.  

I think package-agnostic code is a goal in itself, but there's of course a bit of a conflict of interest with goals like self-contained code and brevity.

**Comment by jameshensman** [comment]
@Bonnevie ,

I agree that there is perhaps a separate need for a kernel toolbox that is language agnostic. But I'm not sure the GPflow is the right home for it. 

For me, (1) is a killer. I have multiple other projects where I have knocked up a fancy kernel with a few extra parameters: the GPflow framework is indispensable there. Even if you were to replace the kernel framework with something as fully featured, It would be frustrating not to have the same methods for extending kernels as other bits of the model. 

Perhaps what's really needed is better methods on kernels for standalone use. You say "kernels will only work inside a Model": exactly which bits don't work? You can certainly call `compute_K(X)` on a kernel without attaching it to a model. 

For integrating with Edward, I think perhaps what's needed is a way to attach the kernels 'free_state' variable to a random variable in Edward. I think this might be as easy as calling make_tf_array(s) on the Edward tensorflow variable.

**Comment by markvdw** [comment]
I actually think that the whole `param.py` in GPflow could be a separate project: A project that deals with parameters you want to optimize over. This is a feature that Edward and GPflow both need. I haven't really had the time to look much at Edward (from the surface it looks cool), so I don't really know how they do this.

As James points out, there may be a hack to get it to work in the short term, but perhaps there are some bigger ideas here?

**Comment by Bonnevie** [comment]
@jameshensman: but if we split off the kernel package and then did a gpflowkernels.py file where we wrapped all functions appropriately, would we really be losing anything? As to your question about what you cannot do at the moment, the main restriction is that it doesn't mesh well with other Tensorflow tensors. `compute_K` is AutoFlow, and can thus only be used in a Numpy setting and there we already have the GPy kernels. To work with Tensorflow you need to activate `tf_mode` appropriately.  

@markvdw: splitting off `param.py` would be quite useful too (and I guess historically the same thing happened with GPy and the Paramz package?). As a bit of blue sky thinking, having a model construction language most practitioners could rally around would be great for the community, but admittedly a pretty ambitious goal. 

There are a couple of Bayes-in-Tensorflow packages floating around, I believe, all with slightly different syntax: GPflow, Edward, PyMC3, BayesFlow in the TF contrib folder, and then I think TF-learn has something called the `BaseEstimator` class which works a bit like your `Model` with some utility built in. But I think the GPflow setup has the highest potential for wide-spread use since it trivializes optimization to such a high degree with both built-in constraints and Scipy integration.

**Comment by Bonnevie** [comment]
Something like this is what I'm suggesting.

```
from kernelflow import AgnosticKern 
#AgnosticKern is the Kern class with no Parameterized inheritance


class Static(AgnosticKern):
    """
    Kernels who don't depend on the value of the inputs are 'Static'.  The only
    parameter is a variance.
    """

    def __init__(self, input_dim, variance=1.0, active_dims=None):
        AgnosticKern.__init__(self, input_dim, active_dims)
        self.variance = variance

    def Kdiag(self, X):
        return tf.fill(tf.stack([tf.shape(X)[0]]), tf.squeeze(self.variance))

#the following class should be identical in every sense to the GPflow kernel.
class ParameterizedStatic(Static, Parameterized): 
    def __init__(self, input_dim, variance=1.0, active_dims=None):
        Parameterized.__init__(self)
        variance = Param(variance, transforms.positive)
        Static.__init__(self, input_dim, variance, active_dims)
    
```

**Comment by junpenglao** [comment]
Hi there, 
I am also looking into combining Edward and GPflow. This might be a naive suggestion as I have no idea of the underlying mechanics of GPflow, but is it possible to get the _graph from GPflow and used it in as part of a larger computation graph (which will give the opportunity to incorporate Edward)?

**Comment by dustinvtran** [comment]
Hello, dropping in again. (I closely monitor GPflow development as I find it a cool package!)

> I actually think that the whole param.py in GPflow could be a separate project: A project that deals with parameters you want to optimize over. This is a feature that Edward and GPflow both need. [...]

In Edward, we try to use as few abstractions as possible and rely strictly on TensorFlow semantics. I think this has a lot of advantages, such as extensibility (more abstractions lead to lengthier contracts and thus more restrictions), readability, and compatibility with other packages. For example, instead of a Parameter object carrying its own graph structure, we look up dependencies from the computational graph (see, e.g., [`get_variables()`](https://github.com/blei-lab/edward/blob/master/edward/util/random_variables.py#L551)). Using the computational graph would solve issue 1 in https://github.com/GPflow/GPflow/issues/322#issuecomment-275369974. Issue 2 on automatic initialization can still be done; it just depends on which class you want the initialization to be run.

As another example, instead of coding up a separate `densities.py` and `transforms.py`, it would be nice to take advantage of the distributions and Bijectors class in `tf.contrib.distributions`. These are more comprehensive in robustness and vectorization. As `tf.contrib.distributions` is moving soon into core TensorFlow, it will continue to be expanded upon as the official distribution (and transformations) library.

There's a lot of cool things that we might be able to do by joining efforts—whether it be on building GPs for any probabilistic model or coding up model-specific inference blocks—and many of us in the Blei Lab would love to collaborate. @junpenglao's comment (https://github.com/GPflow/GPflow/issues/322#issuecomment-277520692) and @alexggmatthews's in another issue get at a key challenge, which would be to better expose the TensorFlow infrastructure in GPflow. Is there anything I might be able to help with?

> For integrating with Edward, I think perhaps what's needed is a way to attach the kernels 'free_state' variable to a random variable in Edward. I think this might be as easy as calling make_tf_array(s) on the Edward tensorflow variable.

I looked into the code base; could you elaborate on this?

**Comment by Bonnevie** [comment]
@dustinvtran: as to your question at the end, `Param` objects are not instantiated as Tensorflow arrays. To integrate appropriately with optimization routines, a single vector-shaped `tf.Variable` is instantiated at the highest model level and each `Param` object is then made to call its method `make_tf_array(arr)` which takes said variable and bites off a slice, reshapes it, and transforms it appropriately. So by the end each `Param` object owns a small slice of this single variable. Although I think there are in-house tools now for integrating with the SciPy optimizers, this approach still provides a convenient way to handle all the necessary transforms. 
What was suggested (I think) was that if the `Param` instead points to a relevant Edward tensor, then everything might still work seamlessly.

**Comment by alexggmatthews** [comment]
@dustinvtran there are some interesting suggestions here thank you. I've separated some of them off into separate issues as this thread is not at risk of becoming too diverse.

**Comment by jameshensman** [comment]
It's a good plan to use tf.distributions instead of densities.py. When we first wrote densities.py, that didn't exist. I looked at using it a while ago but decided that since it was in contrib it was subject to too much flux, so better to keep using densities.py. If we deem tf.distributions mature it would of course make sense to use them.

@dustinvtran I don't see any transformations in tf.contrib.distributions. Unless you count the 'built-in' transformations to some distributions such as "BetaWithSoftplusAB". I'm not sure that's an elegant solution: GPflow allows the user to mix-and-match transforms. 

@Bonnevie your description of 'biting off' chunks of a variable is apt. The nice thing about this set up is that we can plug in to any optimizer you like, or any MCMC code you want. For example, in another in-the-works project I have implemented split HMC for GPflow, which was very simple because of this setup. 

One of the things that the Param class adds is the ability to 'play' with a model instance. This is often necessary to understand what's happening in a complicated model like a GPLVM for instance. For example, you can set parameters to sensible initial values (just using assignment) or declare them 'fixed' (kern.lengthscales.fixed = True) whilst optimizing other parameters. So a one-to-one mapping from tf.variables to GPflow parameters is not quite the correct picture. 

I think the way to integrate with Edward would be to stitch GPflow's `free_state` variable (the long vector that each parameter takes a bite of) into an Edward variable. I'd be curious as to what the perceived benefit would be? It might be interesting to try out Edward's ADVI inference on a GPflow model, perhaps? I'm genuinely curious as to the use-cases, beyond seeming like a fun thing to do.

**Comment by junpenglao** [comment]
> I don't see any transformations in tf.contrib.distributions.  

there is a tf.contrib.distributions.TransformedDistribution in tensorflow.

>  I'd be curious as to what the perceived benefit would be?  

I am interested in using the Edward Variables in a hierarchical bayesian way to inference the kernel parameters in the cov func. Similar to the [example](https://github.com/GPflow/GPflow/blob/master/doc/source/notebooks/mcmc.ipynb) where:
`m.kern.matern32.lengthscales.prior = GPflow.priors.Gamma(1., 1.)`
I would like to do something like:
`m.kern.matern32.lengthscales.prior = X*beta + eps`
which `beta` is a ed.Variable following some prior distribution
(maybe this is also possible to do in GPflow right now? I couldnt find an example for that)

**Comment by Bonnevie** [comment]
@jameshensman: I think Dustin was referring to [this contrib](https://www.tensorflow.org/api_docs/python/contrib.distributions.bijector/) of random variable transfromations, maybe?

**Comment by mathDR** [comment]
Was there any resolution/movement on this? (Tying of parameters?)
